### PR TITLE
Add `by-layers` support to analyze api(s)

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR Adds `by-layers` support across all analyze api(s); the by_layers analysis tragets Linear Operations such as Conv, MatMul, Gemm and all their variants for now.

Tests: manual run with both dense and quantized resnet50 models

- [X] Dense 
```bash
sparsezoo.analyze ~/models/resnet50.onnx \
  --by-types True \
  --by-layers True
```

```bash
INFO:root:Starting Analysis ...
INFO:root:Analysis complete, collating results...
INFO:root:Running analysis `by_layers`
INFO:root:Running analysis `by_types`
Analysis by Layers:
                                    NAME          TOTAL            SIZE        SPARSITY       QUANTIZED 
                                  Conv_0        9408.00          301056            0.00            0.00 
                                  Conv_4        4096.00          131072            0.00            0.00 
                                  Conv_7       36864.00         1179648            0.00            0.00 
                                 Conv_10       16384.00          524288            0.00            0.00 
                                 Conv_12       16384.00          524288            0.00            0.00 
                                 Conv_16       16384.00          524288            0.00            0.00 
                                 Conv_19       36864.00         1179648            0.00            0.00 
                                 Conv_22       16384.00          524288            0.00            0.00 
                                 Conv_26       16384.00          524288            0.00            0.00 
                                 Conv_29       36864.00         1179648            0.00            0.00 
                                 Conv_32       16384.00          524288            0.00            0.00 
                                 Conv_36       32768.00         1048576            0.00            0.00 
                                 Conv_39      147456.00         4718592            0.00            0.00 
                                 Conv_42       65536.00         2097152            0.00            0.00 
                                 Conv_44      131072.00         4194304            0.00            0.00 
                                 Conv_48       65536.00         2097152            0.00            0.00 
                                 Conv_51      147456.00         4718592            0.00            0.00 
                                 Conv_54       65536.00         2097152            0.00            0.00 
                                 Conv_58       65536.00         2097152            0.00            0.00 
                                 Conv_61      147456.00         4718592            0.00            0.00 
                                 Conv_64       65536.00         2097152            0.00            0.00 
                                 Conv_68       65536.00         2097152            0.00            0.00 
                                 Conv_71      147456.00         4718592            0.00            0.00 
                                 Conv_74       65536.00         2097152            0.00            0.00 
                                 Conv_78      131072.00         4194304            0.00            0.00 
                                 Conv_81      589824.00        18874368            0.00            0.00 
                                 Conv_84      262144.00         8388608            0.00            0.00 
                                 Conv_86      524288.00        16777216            0.00            0.00 
                                 Conv_90      262144.00         8388608            0.00            0.00 
                                 Conv_93      589824.00        18874368            0.00            0.00 
                                 Conv_96      262144.00         8388608            0.00            0.00 
                                Conv_100      262144.00         8388608            0.00            0.00 
                                Conv_103      589824.00        18874368            0.00            0.00 
                                Conv_106      262144.00         8388608            0.00            0.00 
                                Conv_110      262144.00         8388608            0.00            0.00 
                                Conv_113      589824.00        18874368            0.00            0.00 
                                Conv_116      262144.00         8388608            0.00            0.00 
                                Conv_120      262144.00         8388608            0.00            0.00 
                                Conv_123      589824.00        18874368            0.00            0.00 
                                Conv_126      262144.00         8388608            0.00            0.00 
                                Conv_130      262144.00         8388608            0.00            0.00 
                                Conv_133      589824.00        18874368            0.00            0.00 
                                Conv_136      262144.00         8388608            0.00            0.00 
                                Conv_140      524288.00        16777216            0.00            0.00 
                                Conv_143     2359296.00        75497472            0.00            0.00 
                                Conv_146     1048576.00        33554432            0.00            0.00 
                                Conv_148     2097152.00        67108864            0.00            0.00 
                                Conv_152     1048576.00        33554432            0.00            0.00 
                                Conv_155     2359296.00        75497472            0.00            0.00 
                                Conv_158     1048576.00        33554432            0.00            0.00 
                                Conv_162     1048576.00        33554432            0.00            0.00 
                                Conv_165     2359296.00        75497472            0.00            0.00 
                                Conv_168     1048576.00        33554432            0.00            0.00 
                                Gemm_179     2049000.00        65568000            0.00            0.00 

Parameters by types:
                                    TYPE           SIZE        SPARSITY       QUANTIZED 
                                  Weight      816093184            0.00            0.00 
                                    Bias          32000            0.00            0.00 
                                   Total      816125184            0.00            0.00 

Ops by types:
                                    TYPE           SIZE        SPARSITY       QUANTIZED 
                                    Gemm       65568000            0.00            0.00 
                                    Conv      750557184            0.00            0.00 
                      BatchNormalization           1696            0.00            0.00 
                                 Softmax             32            0.00            0.00 
                                   Shape             32            0.00            0.00 
                                  Concat             32            0.00            0.00 
                                    Relu           1568            0.00            0.00 
                                  Gather             32            0.00            0.00 
                               Unsqueeze             32            0.00            0.00 
                                 Reshape             32            0.00            0.00 
                       GlobalAveragePool             32            0.00            0.00 
                                     Add            512            0.00            0.00 
                                 MaxPool             32            0.00            0.00 
                                Constant             32            0.00            0.00 
                                   Total      816129248            0.00            0.00 

Params:
                                   MODEL       SPARSITY       QUANTIZED           COUNT            SIZE 
        ~/models/resnet50.onnx           0.00            0.00        25503912       816125184 

Ops:
                                   MODEL       SPARSITY       QUANTIZED           COUNT            SIZE 
        ~/models/resnet50.onnx           0.00            0.00        25504039       816129248 

Overall:
                                   MODEL       SPARSITY       QUANTIZED 
        ~/models/resnet50.onnx           0.00            0.00 

```

- [X] Quantized 
```bash
sparsezoo.analyze ~/models/resnet50-quac.onnx \
  --by-types True \
  --by-layers True
```

```bash
INFO:root:Starting Analysis ...
INFO:root:Analysis complete, collating results...
INFO:root:Running analysis `by_layers`
INFO:root:Running analysis `by_types`
Analysis by Layers:
                                    NAME          TOTAL            SIZE        SPARSITY       QUANTIZED 
                           Conv_13_quant        9472.00           77312           47.12           99.32 
                           Conv_37_quant        4160.00           34816           57.93           98.46 
                          Conv_105_quant       16640.00          139264           73.60           98.46 
                           Conv_60_quant       36928.00          296960           82.83           99.83 
                           Conv_83_quant       16640.00          139264           76.02           98.46 
                          Conv_129_quant       16448.00          133120           76.36           99.61 
                          Conv_152_quant       36928.00          296960           79.90           99.83 
                          Conv_175_quant       16640.00          139264           77.71           98.46 
                          Conv_199_quant       16448.00          133120           73.05           99.61 
                          Conv_222_quant       36928.00          296960           73.94           99.83 
                          Conv_245_quant       16640.00          139264           75.97           98.46 
                          Conv_269_quant       32896.00          266240           69.00           99.61 
                          Conv_337_quant      131584.00         1064960           91.45           99.61 
                          Conv_292_quant      147584.00         1183744           86.41           99.91 
                          Conv_315_quant       66048.00          540672           83.29           99.22 
                          Conv_361_quant       65664.00          528384           93.73           99.81 
                          Conv_384_quant      147584.00         1183744           93.65           99.91 
                          Conv_407_quant       66048.00          540672           91.26           99.22 
                          Conv_431_quant       65664.00          528384           86.88           99.81 
                          Conv_454_quant      147584.00         1183744           89.82           99.91 
                          Conv_477_quant       66048.00          540672           84.39           99.22 
                          Conv_501_quant       65664.00          528384           85.17           99.81 
                          Conv_524_quant      147584.00         1183744           86.82           99.91 
                          Conv_547_quant       66048.00          540672           86.34           99.22 
                          Conv_571_quant      131328.00         1056768           75.72           99.81 
                          Conv_639_quant      525312.00         4227072           95.26           99.81 
                          Conv_594_quant      590080.00         4726784           94.36           99.96 
                          Conv_617_quant      263168.00         2129920           85.98           99.61 
                          Conv_663_quant      262400.00         2105344           97.03           99.90 
                          Conv_686_quant      590080.00         4726784           97.26           99.96 
                          Conv_709_quant      263168.00         2129920           93.42           99.61 
                          Conv_733_quant      262400.00         2105344           95.61           99.90 
                          Conv_756_quant      590080.00         4726784           96.79           99.96 
                          Conv_779_quant      263168.00         2129920           92.10           99.61 
                          Conv_803_quant      262400.00         2105344           94.13           99.90 
                          Conv_826_quant      590080.00         4726784           96.37           99.96 
                          Conv_849_quant      263168.00         2129920           92.91           99.61 
                          Conv_873_quant      262400.00         2105344           93.77           99.90 
                          Conv_896_quant      590080.00         4726784           97.29           99.96 
                          Conv_919_quant      263168.00         2129920           94.49           99.61 
                          Conv_943_quant      262400.00         2105344           92.50           99.90 
                          Conv_966_quant      590080.00         4726784           96.95           99.96 
                          Conv_989_quant      263168.00         2129920           93.72           99.61 
                         Conv_1013_quant      524800.00         4210688           81.01           99.90 
                         Conv_1081_quant     2099200.00        16842752           98.09           99.90 
                         Conv_1036_quant     2359808.00        18890752           99.35           99.98 
                         Conv_1059_quant     1050624.00         8454144           93.24           99.81 
                         Conv_1105_quant     1049088.00         8404992           97.91           99.95 
                         Conv_1128_quant     2359808.00        18890752           98.61           99.98 
                         Conv_1151_quant     1050624.00         8454144           96.63           99.81 
                         Conv_1175_quant     1049088.00         8404992           92.14           99.95 
                         Conv_1198_quant     2359808.00        18890752           99.51           99.98 
                         Conv_1221_quant     1050624.00         8454144           94.30           99.81 
                               Gemm_1239     2049000.00        65568000           87.42            0.00 

Parameters by types:
                                    TYPE           SIZE        SPARSITY       QUANTIZED 
                                  Weight      253175296           95.00           91.97 
                                    Bias         881920            0.00            0.00 
                                   Total      254057216           94.90           91.87 

Ops by types:
                                    TYPE           SIZE        SPARSITY       QUANTIZED 
                             QLinearConv      187851776           95.55          100.00 
                                    Gemm       65568000           87.42            0.00 
                          QuantizeLinear            544            0.00            0.00 
                       GlobalAveragePool             32            0.00            0.00 
                                 Softmax             32            0.00            0.00 
                                 Reshape             32            0.00            0.00 
                                  Concat             32            0.00            0.00 
                               Unsqueeze             32            0.00            0.00 
                                   Shape             32            0.00            0.00 
                                     Add            512            0.00            0.00 
                                    Relu            416            0.00            0.00 
                        DequantizeLinear           1056            0.00            0.00 
                                  Gather             32            0.00            0.00 
                                 MaxPool             32            0.00            0.00 
                                   Total      253422560           94.90           91.97 

Params:
                                   MODEL       SPARSITY       QUANTIZED           COUNT            SIZE 
   /home/rahul/models/resnet50-quac.onnx          94.90           91.87        25530472       254057216 

Ops:
                                   MODEL       SPARSITY       QUANTIZED           COUNT            SIZE 
   /home/rahul/models/resnet50-quac.onnx          94.90           91.97        25530559       253422560 

Overall:
                                   MODEL       SPARSITY       QUANTIZED 
   /home/rahul/models/resnet50-quac.onnx          94.90           91.92 

```

### CODE REVIEW

**Review by dbogunowicz — APPROVED** [review]

File: src/sparsezoo/analyze/analysis.py [hunk 1]
```diff
@@ -660,12 +672,54 @@ def from_model_analysis(
...
+            by_layers_entries = []
+            for node in analysis.nodes:
+                if node.op_type not in LINEAR_OP_TYPES:
+                    continue
+                precision_dict = node.parameter_summary.precision
+                dense = 0
+                quantized = 0
+                for precision, counts in precision_dict.items():
+                    if "32" in precision:
+                        # include float32, int32
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparsezoo/analyze/analysis.py:691]: Right now the analyze api only supports fp32, int32 and int8 precision; but will be adding support for different precisions in future, hence the TODO

File: src/sparsezoo/analyze/analysis.py [hunk 2]
```diff
@@ -660,12 +672,54 @@ def from_model_analysis(
...
         """
         sections = []
 
         if by_layers:
-            # TODO: Add analysis by_layers section
-            _LOGGER.info("analysis `by_layer` is not implemented yet, will be ignored")
+            _LOGGER.info("Running analysis `by_layers`")
+            by_layers_entries = []
+            for node in analysis.nodes:
+                if node.op_type not in LINEAR_OP_TYPES:
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/sparsezoo/analyze/analysis.py]: could you elaborate?

[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparsezoo/analyze/analysis.py]: Syncing offline!

**Review by KSGulin — APPROVED** [review]
Code looks good! But as discussed offline, the biases are getting quantized to int32 and are not being counted in the quantized weight total % for each layer, which is reasonable. However, in the `Parameters by types` section, it's shown that ~96% of biases are quantized, which I would expect to be 0. I think this is worth investigating, but it's not directly tied to this PR, so I'm approving

**Comment by rahul-tuli** [comment]
> Code looks good! But as discussed offline, the biases are getting quantized to int32 and are not being counted in the quantized weight total % for each layer, which is reasonable. However, in the `Parameters by types` section, it's shown that ~96% of biases are quantized, which I would expect to be 0. I think this is worth investigating, but it's not directly tied to this PR, so I'm approving

What a great Catch, you were correct, this was a bug that stemmed from a mismatch where in by-layers analysis fp32, int32 were considered dense and everything else quantized, whereas in by-types analysis only fp32 was considered dense, I've fixed this in latest commit!

**Pull request merged**